connectors/sources/servicenow.py

```diff
@@ -0,0 +1,346 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""ServiceNow source module responsible to fetch documents from ServiceNow."""
+import json
+import os
+
+import aiohttp
+from aiohttp.client_exceptions import ServerDisconnectedError
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import CancellableSleeps, iso_utc
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+FETCH_SIZE = 1000
+MAX_CONCURRENCY_SUPPORT = 10
+
+RUNNING_FTEST = (
+    "RUNNING_FTEST" in os.environ
+)  # Flag to check if a connector is run for ftest or not.
+
+ENDPOINTS = {
+    "TABLE": "/api/now/table/{table}",
+    "ATTACHMENT": "/api/now/attachment",
+    "DOWNLOAD": "/api/now/attachment/{sys_id}/file",
+}
+
+
+class InvalidResponse(Exception):
+    """Raise when retrieving invalid response..
+
+    Args:
+        Exception (InvalidResponse): Invalid response exception.
+    """
+
+    pass
+
+
+class ServiceNowClient(BaseDataSource):
+    """ServiceNow Client"""
+
+    def __init__(self, configuration):
+        """Setup the ServiceNow client.
+
+        Args:
+            configuration (DataSourceConfiguration): Instance of DataSourceConfiguration class.
+        """
+
+        self.session = None
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self.services = self.configuration["services"]
+        self.retry_count = self.configuration["retry_count"]
+
+    def _get_session(self):
+        """Generate aiohttp client session with configuration fields.
+
+        Returns:
+            aiohttp.ClientSession: An instance of Client Session
+        """
+
+        if self.session:
+            return self.session
+
+        logger.debug("Generating aiohttp client session")
+        connector = aiohttp.TCPConnector(limit=MAX_CONCURRENCY_SUPPORT)
```

Comment:
Question: Should we use the configuration value `concurrent_downloads` here? Or is this something else?

Reply:
For not limiting client creation in the session

Reply:
Right, but in the configuration there's a field called `concurrent_downloads`. Shouldn't it be used here instead of the const `MAX_CONCURRENCY_SUPPORT`?

Reply:
If user enters `concurrent_downloads` value as 5 then it will restrict client creation in session only with 5 which may affect performance. Our source handles 10 clients at a time

Reply:
Okay, so it's a different use-case for the `concurrent_downloads` configuration field. What is `concurrent_downloads` used for then, if not this? Is it being added later?

Reply:
Yup, It will be used for handling concurrency for downloading attachment content